clients/pkg/promtail/targets/file/decompresser.go

```diff
@@ -0,0 +1,285 @@
+	"github.com/prometheus/common/model"
+	"go.uber.org/atomic"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/ianaindex"
+	"golang.org/x/text/transform"
+
+	"github.com/grafana/loki/pkg/logproto"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/positions"
+)
+
+type decompresser struct {
+	metrics   *Metrics
+	logger    log.Logger
+	handler   api.EntryHandler
+	positions positions.Positions
+
+	path string
+
+	posAndSizeMtx sync.Mutex
+	stopOnce      sync.Once
+
+	running *atomic.Bool
+	posquit chan struct{}
+	posdone chan struct{}
+	done    chan struct{}
+
+	decoder *encoding.Decoder
+
+	position int64
+	size     int64
+}
+
+func newDecompresser(metrics *Metrics, logger log.Logger, handler api.EntryHandler, positions positions.Positions, path string, encodingFormat string) (*decompresser, error) {
+	logger = log.With(logger, "component", "decompresser")
+
+	pos, err := positions.Get(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "get positions")
+	}
+
+	var decoder *encoding.Decoder
+	if encodingFormat != "" {
+		level.Info(logger).Log("msg", "decompresser will decode messages", "from", encodingFormat, "to", "UTF8")
+		encoder, err := ianaindex.IANA.Encoding(encodingFormat)
+		if err != nil {
+			return nil, errors.Wrap(err, "error doing IANA encoding")
+		}
+		decoder = encoder.NewDecoder()
+	}
+
+	decompresser := &decompresser{
+		metrics:   metrics,
+		logger:    logger,
+		handler:   api.AddLabelsMiddleware(model.LabelSet{FilenameLabel: model.LabelValue(path)}).Wrap(handler),
+		positions: positions,
+		path:      path,
+		running:   atomic.NewBool(false),
+		posquit:   make(chan struct{}),
+		posdone:   make(chan struct{}),
+		done:      make(chan struct{}),
+		position:  pos,
+		decoder:   decoder,
+	}
+
+	go decompresser.readLines()
+	go decompresser.updatePosition()
+	metrics.filesActive.Add(1.)
+	return decompresser, nil
+}
+
+// mountReader instantiate a reader ready to be used by the decompresser.
+//
+// The selected reader implementation is based on the extension of the given file name.
+// It'll error if the extension isn't supported.
+func mountReader(f *os.File, logger log.Logger) (reader io.Reader, err error) {
+	ext := filepath.Ext(f.Name())
+	var decompressLib string
+
+	if strings.Contains(ext, "gz") { // .gz, .tar.gz
+		decompressLib = "compress/gzip"
+		reader, err = gzip.NewReader(f)
+	} else if ext == ".z" {
+		decompressLib = "compress/zlib"
+		reader, err = zlib.NewReader(f)
+	} else if ext == ".zip" {
+		decompressLib = "compress/flate"
+		reader = flate.NewReader(f)
+	} else if ext == ".bz2" {
+		decompressLib = "bzip2"
+		reader = bzip2.NewReader(f)
+	}
+
+	level.Info(logger).Log("msg", fmt.Sprintf("using %q to decompress file %q", decompressLib, f.Name()))
+
+	if reader != nil {
+		return reader, nil
+	}
+
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	return nil, fmt.Errorf("file %q with unsupported extension", f.Name())
```

Comment:
```suggestion
	return nil, fmt.Errorf("file %q has unsupported extension", f.Name())
```

Might be nice to include which extensions are actually supported in this message

Reply:
Agreed. I included them on https://github.com/grafana/loki/pull/6708/commits/e80a2f87d6da59469d0b84c1588b0d1d7ca685d4.